File: javacommons/gcfprotocols/file/tsrc/mt/javasrc/com/nokia/mj/test/file/utils/FileTestUtils.java
/*
* Copyright (c) 2008 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description:
*
*/


package com.nokia.mj.test.file.utils;

import java.io.IOException;
import javax.microedition.io.file.*;
import javax.microedition.io.*;

public class FileTestUtils
{
    public static void IllegalArgumentExceptionCheck(String aUri,
            boolean aIsSuccessCase) throws IOException
    {
        FileConnection fileConnection = null;
        try
        {
            fileConnection = (FileConnection) Connector.open(aUri);
            if (aIsSuccessCase)
            {
                fileConnection.close();
                return;
            }
            throw new TestFailedException();
        }
        catch (IllegalArgumentException ex)
        {
            if (aIsSuccessCase)
                throw new TestFailedException(ex.toString());
        }
        catch (Exception ex)
        {
            if (fileConnection != null)
                fileConnection.close();
            // append(ex.toString());
            ex.printStackTrace();
            throw new TestFailedException(ex.toString());
        }
    }

    public static void createObject(String aUri, boolean aIsFile)
    {
        FileConnection fileConnection = null;
        try
        {
            fileConnection = (FileConnection) Connector.open(aUri);
            if (aIsFile)
                fileConnection.create();
            else
                fileConnection.mkdir();
        }
        catch (Exception ex)
        {
            // append(ex.toString());
            ex.printStackTrace();
            throw new TestFailedException(ex.toString());
        }
        finally
        {
            if (fileConnection != null)
            {
                try
                {
                    fileConnection.close();
                }
                catch (Exception e)
                {
                    throw new TestFailedException(e.toString());
                }
            }
        }// end finally
    }

    public static void endpointExists(String aUri)
    {
        FileConnection fileConnection = null;
        try
        {
            fileConnection = (FileConnection) Connector.open(aUri);
            boolean retValue = fileConnection.exists();
            // fileConnection.close();
            if (false == retValue)
                throw new TestFailedException();
        }
        catch (Exception ex)
        {
            // append(ex.toString());
            ex.printStackTrace();
            throw new TestFailedException(ex.toString());
        }
        finally
        {
            if (fileConnection != null)
            {
                try
                {
                    fileConnection.close();
                }
                catch (Exception e)
                {
                    throw new TestFailedException();
                }
            }
        }// end finally
    }

    public static void IoExceptionCheck(String aUri, int aMode,
                                        boolean aNoException)
    {
        FileConnection fileConnection = null;
        try
        {
            fileConnection = (FileConnection) Connector.open(aUri, aMode);
            if (aNoException == false)
                throw new TestFailedException();
        }
        catch (IOException ex)
        {
            ex.printStackTrace();
            if (aNoException)
                throw new TestFailedException("#1 " + ex.toString());
        }
        catch (Exception ex)
        {
            // append(ex.toString());
            ex.printStackTrace();
            throw new TestFailedException("#2 " + ex.toString());
        }
        finally
        {
            if (fileConnection != null)
            {
                try
                {
                    fileConnection.close();
                }
                catch (Exception e)
                {
                    throw new TestFailedException("#3 " + e.toString());
                }
            }
        }// end finally
    }

    public static FileConnection openExistsURI(String aUri, int aMode)
    {
        FileConnection fileConnection = null;
        try
        {
            fileConnection = (FileConnection) Connector.open(aUri, aMode);
            if (fileConnection.exists())
            {
                return fileConnection;
            }
            else
            {
                fileConnection.close();
                return null;
            }
        }
        catch (Exception ex)
        {
            // append(ex.toString());
            ex.printStackTrace();
            return null;
        }
    }

    public static void testGetConnectionMethods(String aEndpoint, String aName,
            String aPath, String aUri) throws IOException
    {
        FileConnection conn1 = null;
        conn1 = (FileConnection) Connector.open(aEndpoint, Connector.READ);
        String name = conn1.getName();
        if (false == name.equals(aName))
            throw new TestFailedException("Name: " + name + " Expected: "
                                          + aName);
        String path = conn1.getPath();
        if (false == path.equals(aPath))
            throw new TestFailedException("path: " + path + " Expected: "
                                          + aPath);
        String uri = conn1.getURL();
        if (false == uri.equals(aUri))
            throw new TestFailedException("uri: " + uri + " Expected: " + aUri);
        conn1.close();
    }

    public static void createFile(String aUri) throws IOException
    {
        FileConnection conn1 = (FileConnection) Connector.open(aUri,
                               Connector.READ_WRITE);
        if (conn1.exists())
            throw new TestFailedException();
        conn1.create();
        conn1.close();
        // Re-open a file to confirm that file was created on the filesystem.
        conn1 = openExistsURI(aUri, Connector.READ);
        if (conn1 == null)
            throw new TestFailedException();
        if (true == conn1.isDirectory())
            throw new TestFailedException();
        conn1.close();
    }

    public static void createDir(String aUri) throws IOException
    {
        FileConnection conn1 = (FileConnection) Connector.open(aUri,
                               Connector.READ_WRITE);
        if (conn1.exists())
            throw new TestFailedException();
        conn1.mkdir();
        conn1.close();
        // Re-open a file to confirm that file was created on the filesystem.
        conn1 = openExistsURI(aUri, Connector.READ);
        if (conn1 == null)
            throw new TestFailedException();
        if (false == conn1.isDirectory())
            throw new TestFailedException();
        conn1.close();
    }

    public static void illegalArgumentFromRenaming(String aNewName, String aUri)
    throws IOException
    {
        FileConnection conn1 = null;
        boolean exceptionOccured = false;
        conn1 = (FileConnection) Connector.open(aUri, Connector.READ_WRITE);
        if (!conn1.exists())
            throw new TestFailedException();
        exceptionOccured = false;
        try
        {
            conn1.rename(aNewName);
        }
        catch (IllegalArgumentException ex)
        {
            exceptionOccured = true;
        }
        if (exceptionOccured == false)
            throw new TestFailedException();
        conn1.close();
    }

    public static void IOExceptionFromRenaming(String aNewName, String aUri)
    throws IOException
    {
        FileConnection conn1 = null;
        boolean exceptionOccured = false;
        conn1 = (FileConnection) Connector.open(aUri, Connector.READ_WRITE);
        if (!conn1.exists())
            throw new TestFailedException();
        exceptionOccured = false;
        try
        {
            conn1.rename(aNewName);
        }
        catch (IOException ex)
        {
            exceptionOccured = true;
        }
        if (exceptionOccured == false)
            throw new TestFailedException();
        conn1.close();
    }
}

File: javacommons/gcfprotocols/file/tsrc/mt/filetestinit/src.linux/filetestutils.cpp
/*
* Copyright (c) 2008 Nokia Corporation and/or its subsidiary(-ies).
* All rights reserved.
* This component and the accompanying materials are made available
* under the terms of "Eclipse Public License v1.0"
* which accompanies this distribution, and is available
* at the URL "http://www.eclipse.org/legal/epl-v10.html".
*
* Initial Contributors:
* Nokia Corporation - initial contribution.
*
* Contributors:
*
* Description:  FileTestUtils
 *
*/


#include <fstream>
#include <iostream>
#include <stdio.h>
#include <errno.h>
#include "filetestutils.h"

using namespace std;
using namespace java::filetest;

void FileTestUtils::createFile(const wstring &aPath, const int aFlags, int aMode)
{
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());

    if (0 == aMode)
    {
        aMode = 0777;
    }
    int fd = open(chr, aFlags, aMode);
    FileTestUtils::checkError(aPath,fd);
    close(fd);
}


void FileTestUtils::createSystemFile(const wstring &aPath)
{
    FileTestUtils::createFile(aPath, O_CREAT|O_EXCL|O_WRONLY);
    //There is nothing like System File in Linux. So do nothing.
}

void FileTestUtils::createHiddenDir(const wstring &aPath, const wstring& aFile)
{
    wstring name = aPath+L"."+aFile;
    FileTestUtils::createDir(name.c_str());
    //Creating dir which starts with .is as good as hidden :-)
}

void FileTestUtils::createSystemDir(const wstring &aPath)
{
    FileTestUtils::createDirMode(aPath, 0777);
    //There is nothing like System File in Linux. So do nothing.
}

void FileTestUtils::createDirMode(const wstring& aPath, const int aMode)
{
    // R W X
    // 4 2 1
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    int error = mkdir(chr, aMode);
    FileTestUtils::checkError(aPath, error);
}

void FileTestUtils::writeDataToFile(const wstring& aPath, const string& aData)
{
    //Linux, just use ofstream
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    int fd = open(chr, O_RDWR);
    CheckError(aPath, fd);
    write(fd, aData.c_str(), aData.length());
    close(fd);
}

void FileTestUtils::writeDataInLoop(const wstring& aPath, const int aData, const int aTimes)
{
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    ofstream file(chr, ios::out|ios::app|ios::binary);

    for (int i=0 ; i<aTimes; i++)
    {
        file<<(char)aData;
    }
    file.close();

}

void FileTestUtils::appendDataToFile(const wstring& aPath, const int aData)
{
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    ofstream file(chr, ios::out|ios::app|ios::binary);
    file<<(char)aData;
    file.close();
}

void FileTestUtils::appendDataToFile(const wstring& aPath, const wstring& aData)
{
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    ofstream file(chr, ios::out|ios::app|ios::binary);
    file<<(char*)aData.c_str();
    file.close();
}

void FileTestUtils::setFileHidden(const wstring &aPath, const wstring &aFile)
{
    //File Starting from . is considered hidden
    wstring oldFile = aPath+aFile;
    wstring newName = aPath+L"."+aFile;

    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());

    char* chr1 = new char[aFile.size()+1];
    sprintf(chr1,"%ls",aFile.c_str());
    rename(chr,chr1);
}

void FileTestUtils::setFileReadOnly(const wstring &aPath)
{
    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());
    chmod(chr, 0444);
}

void FileTestUtils::setDirHidden(const wstring &aPath, const wstring &aName)
{
    wstring oldFile = aPath+aName;
    wstring newName = aPath+L"."+aName;

    char* chr = new char[aPath.size()+1];
    sprintf(chr,"%ls",aPath.c_str());

    char* chr1 = new char[aName.size()+1];
    sprintf(chr1,"%ls",aName.c_str());
    rename(chr,chr1);
}
